apps/rewards/src/invite/rewardSender.service.ts

```diff
@@ -0,0 +1,156 @@
+import { KomenciLoggerService } from '@app/komenci-logger'
+import { ContractKit } from '@celo/contractkit'
+import { Inject, Injectable } from '@nestjs/common'
+import { Cron, CronExpression } from '@nestjs/schedule'
+import BigNumber from 'bignumber.js'
+import { EntityManager, In, Raw, Repository } from 'typeorm'
+import { appConfig, AppConfig } from '../config/app.config'
+import { RelayerProxyService } from '../relayer/relayer_proxy.service'
+import { InviteReward, RewardStatus } from './inviteReward.entity'
+import { InviteRewardRepository } from './inviteReward.repository'
+
+const WEI_PER_UNIT = 1000000000000000000
+
+interface WatchedInvite {
+  inviteId: string
+  txHash: string
+  sentAt: number
+  markAsDeadLetterIfFailed: boolean
+}
+
+@Injectable()
+export class RewardSenderService {
+  watchedInvites: Set<WatchedInvite>
+
+  constructor(
+    private readonly inviteRewardRepository: InviteRewardRepository,
+    private readonly relayerProxyService: RelayerProxyService,
+    @Inject(appConfig.KEY)
+    private readonly appCfg: AppConfig,
+    private readonly contractKit: ContractKit,
+    private readonly logger: KomenciLoggerService
+  ) {
+    this.watchedInvites = new Set()
+  }
+
+  async wasTxSentSuccesfully(txHash: string) {
+    const tx = await this.contractKit.web3.eth.getTransaction(txHash)
+    return Boolean(tx.blockHash)
+  }
+
+  @Cron(CronExpression.EVERY_SECOND)
+  async checkWatchedInvites() {
+    await Promise.all(
+      [...this.watchedInvites].map(async invite => {
+        if (await this.wasTxSentSuccesfully(invite.txHash)) {
+          await this.inviteRewardRepository.update(invite.inviteId, {
+            state: RewardStatus.Completed
+          })
+          this.watchedInvites.delete(invite)
+        } else if (
+          Date.now() - invite.sentAt >
+          this.appCfg.transactionTimeoutMs
+        ) {
+          await this.inviteRewardRepository.update(invite.inviteId, {
+            state: invite.markAsDeadLetterIfFailed
+              ? RewardStatus.DeadLettered
+              : RewardStatus.Failed
+          })
+          this.watchedInvites.delete(invite)
+        }
+      })
+    )
+  }
+
+  @Cron(CronExpression.EVERY_MINUTE)
+  async checkFailedInvites() {
+    await this.inviteRewardRepository.manager.transaction(
+      async (entityManager: EntityManager) => {
+        const repository = entityManager.getRepository<InviteReward>(
+          'invite_reward'
+        )
+        const invites = await repository
+          .createQueryBuilder()
+          .setLock('pessimistic_read')
```

Comment:
This should be `pessimistic_write`, correct? We don't want two reward service instances reading this entity and deciding to retry the payment